xdebug_com.c

```diff
@@ -21,169 +22,242 @@
 #ifdef WIN32
-	ioctlsocket(sockfd, FIONBIO, (u_long*)&yes);
+		ioctlsocket(sockfd, FIONBIO, (u_long*)&yes);
 #else
-	fcntl(sockfd, F_SETFL, O_NONBLOCK);
+		fcntl(sockfd, F_SETFL, O_NONBLOCK);
 #endif
 
-	/* connect */
-	status = connect(sockfd, (struct sockaddr*)&address, sizeof(address));
-	if (status < 0) {
+		/* Try to connect to the newly created socket */
+		/* Worth noting is that the port is set in the getaddrinfo call before */
+		status = connect(sockfd, ptr->ai_addr, ptr->ai_addrlen);
+		
+		/* Determine if we got a connection. If no connection could be made
+		 * we close the socket and continue with the next IP address in the list */
+		if (status < 0) {
 #ifdef WIN32
-		errno = WSAGetLastError();
-		if (errno != WSAEINPROGRESS && errno != WSAEWOULDBLOCK) {
-			close(sockfd);
-			return -1;
-		}
+			errno = WSAGetLastError();
+			if (errno != WSAEINPROGRESS && errno != WSAEWOULDBLOCK) {
+# ifndef DEBUGGER_FAIL_SILENTLY
+				printf("xdebug_create_socket(\"%s\", %d) connect: %d\n",
+						   hostname, dport, errno);
+# endif
 #else
-		if (errno == EACCES) {
-			close(sockfd);
-			return -3;
-		}
-		if (errno != EINPROGRESS) {
-			close(sockfd);
-			return -1;
-		}
+			if (errno == EACCES) {
+# ifndef DEBUGGER_FAIL_SILENTLY
+				printf("xdebug_create_socket(\"%s\", %d) connect: %s\n",
+						   hostname, dport, strerror(errno));
+# endif
+				SCLOSE(sockfd);
+				sockfd = SOCK_ACCESS_ERR;
+				
+				continue;
+			}
+			if (errno != EINPROGRESS) {
+# ifndef DEBUGGER_FAIL_SILENTLY
+				printf("xdebug_create_socket(\"%s\", %d) connect: %s\n",
+						   hostname, dport, strerror(errno));
+# endif
 #endif
-
-		while (1) {
-			fd_set rset, wset, eset;
-
-			FD_ZERO(&rset);
-			FD_SET(sockfd, &rset);
-			FD_ZERO(&wset);
-			FD_SET(sockfd, &wset);
-			FD_ZERO(&eset);
-			FD_SET(sockfd, &eset);
-
-			if (select(sockfd+1, &rset, &wset, &eset, &timeout) == 0) {
-				close(sockfd);
-				return -2;
+				SCLOSE(sockfd);
+				sockfd = SOCK_ERR;
+				
+				continue;
 			}
 
-			/* if our descriptor has an error */
-			if (FD_ISSET(sockfd, &eset)) {
-				close(sockfd);
-				return -1;
+			ufds[0].fd = sockfd;
+			ufds[0].events = POLLIN | POLLOUT | POLLPRI;
+			while (1) {
+				sockerror = poll(ufds, 1, timeout);
+				
+				/* If an error occured when doing the poll */
+				if (sockerror == SOCK_ERR) {
+#ifndef DEBUGGER_FAIL_SILENTLY
+# if WIN32|WINNT
+					printf("create_debugger_socket(\"%s\", %d) WSAPoll: %d\n",
+								hostname, dport, WSAGetLastError());
+# else
+					printf("create_debugger_socket(\"%s\", %d) poll: %s\n",
+								hostname, dport, strerror(errno));
+# endif
+#endif
+					sockerror = SOCK_ERR;
+					break;
+				}
+				
+				/* A timeout occured when polling the socket */
+				if (sockerror == 0) {
+					sockerror = SOCK_TIMEOUT_ERR;
+					break;
+				}
+				
+				/* If the poll was successful but an error occured */
+				if (ufds[0].revents & (POLLERR | POLLHUP | POLLNVAL | POLLRDHUP)) {
```

Comment:
I haven't tried yet (Windows is a pain), but https://www.mail-archive.com/mosquitto-devel@lists.launchpad.net/msg00059.html indicates it won't work on Windows either. So I am inclined to just remove POLLRDHUP.